firewood/src/merkle/proof.rs

```diff
@@ -350,103 +345,118 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
-                                .ok_or(ProofError::ProofNodeMissing)?;
+                Some(p) if key_nibbles.size_hint().0 != 0 => {
+                    let SubProof::Hash(hash) = p else {
+                        // the trie doesn't contain the key
+                        return if allow_non_existent_node {
+                            Ok(None)
+                        } else {
+                            Err(ProofError::NodeNotInTrie)
+                        };
+                    };
+
+                    // If the new parent is a branch node, record the index to correctly link the next child to it.
+                    if node_ref.inner().as_branch().is_some() {
+                        let Some(index) = key_nibbles.next().map(|i| i as usize) else {
+                            return Err(ProofError::NodeNotInTrie);
+                        };
 
-                            chd_ptr = self.decode_node(merkle, cur_key, proof.as_ref(), true)?.0;
+                        child_index = index;
+                    }
 
-                            // Link the child to the parent based on the node type.
-                            match &u_ref.inner() {
-                                // TODO: add path
-                                #[allow(clippy::indexing_slicing)]
-                                NodeType::Branch(n) if n.chd()[branch_index].is_none() => {
-                                    // insert the leaf to the empty slot
-                                    #[allow(clippy::unwrap_used)]
-                                    u_ref
-                                        .write(|u| {
-                                            let uu = u.inner_mut().as_branch_mut().unwrap();
-                                            #[allow(clippy::indexing_slicing)]
-                                            (uu.chd_mut()[branch_index] = Some(chd_ptr));
-                                        })
-                                        .unwrap();
-                                }
+                    child_hash = hash;
+                }
 
-                                // If the child already resolved, then use the existing node.
-                                NodeType::Branch(_) => {}
+                // Return when reaching the end of the key.
+                Some(p) => {
+                    let mut data = None;
+
+                    // Decode the last subproof to get the value.
+                    let node_ref = if let SubProof::Hash(p_hash) = p {
+                        let proof = proofs_map
+                            .get(&p_hash)
+                            .ok_or(ProofError::ProofNodeMissing)?;
+
+                        // If the new parent is a branch node, record the index to correctly link the next child to it.
+                        if node_ref.inner().as_branch().is_some() {
+                            let Some(index) = key_nibbles.next().map(|i| i as usize) else {
```

Comment:
It seems like if we get here, we know that `key_nibbles.size_hint().0 == 0`, otherwise we would have matched the other match statement. I think this means the `next` method is guaranteed to return `None`. If that's not true, can you explain why?